lib/elixir/lib/enum.ex

```diff
@@ -1547,12 +1547,37 @@ defmodule Enum do
 
   ## Examples
 
-      iex> Enum.split_with([1, 2, 3], fn(x) -> rem(x, 2) == 0 end)
-      {[2], [1, 3]}
+      iex> Enum.split_with([5, 4, 3, 2, 1, 0], fn(x) -> rem(x, 2) == 0 end)
+      {[4, 2, 0], [5, 3, 1]}
+
+      iex> Enum.split_with(%{a: 1, b: -2, c: 1, d: -3}, fn({_k, v}) -> v < 0 end)
+      {[b: -2, d: -3], [a: 1, c: 1]}
+
+      iex(3)> Enum.split_with(%{a: 1, b: -2, c: 1, d: -3}, fn({_k, v}) -> v > 50 end)
+      {[], [a: 1, b: -2, c: 1, d: -3]}
+
+      iex(3)> Enum.split_with(%{}, fn({_k, v}) -> v > 50 end)
+      {[], []}
 
   """
   @spec split_with(t, (element -> any)) :: {list, list}
-  def split_with(enumerable, fun) when is_function(fun, 1) do
+  def split_with(enumerable, fun) when is_list(enumerable) and is_function(fun, 1),
+    do: do_split_with_list(enumerable, fun, {[], []})
+  def split_with(enumerable, fun) when is_function(fun, 1),
+    do: do_split_with_enumerable(enumerable, fun)
+
+  defp do_split_with_list([], _fun, {acc1, acc2}),
```

Comment:
It will be faster because we won't need to allocate the tuple but I would find it clearer to. No need to keep it as a tuple if you never return or use it as a tuple?